adapters/improvedigital/improvedigital.go

```diff
@@ -144,3 +148,87 @@ func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType,
 		Message: fmt.Sprintf("Failed to find impression for ID: \"%s\"", impID),
 	}
 }
+
+// This method responsible to check/validate/apply additional consent providers
+// If consent provider check/validate/apply failed, this will return original request
+func (a *ImprovedigitalAdapter) applyAdditionalConsentProviders(request openrtb2.BidRequest) openrtb2.BidRequest {
+	// If user/user.ext not defined, no need to parse additional consent
+	if request.User == nil || request.User.Ext == nil {
+		return request
+	}
+
+	var (
+		err     error
+		reqJSON json.RawMessage
+		cpStr   string
+	)
+
+	// Start validating additional consent
+	// Check key exist user.ext.ConsentedProvidersSettings.consented_providers
+	var userExtMap = make(map[string]json.RawMessage)
+	if err = json.Unmarshal(request.User.Ext, &userExtMap); err != nil {
+		return request
+	}
+
+	cpsMapValue, cpsJSONFound := userExtMap["ConsentedProvidersSettings"]
+	if !cpsJSONFound {
+		return request
+	}
+
+	// Check key exist user.ext.ConsentedProvidersSettings.consented_providers
+	var cpMap = make(map[string]json.RawMessage)
+	if err = json.Unmarshal(cpsMapValue, &cpMap); err != nil {
+		return request
+	}
+
+	cpMapValue, cpJSONFound := cpMap["consented_providers"]
+	if !cpJSONFound {
+		return request
+	}
+	// End validating additional consent
+
+	// Check if string contain ~, then return only numbers
+	if cpStr, err = hasAdditionalConsent(string(cpMapValue)); err != nil {
+		return request
+	}
+
+	// Prepare consent providers string for append
+	cpStr = fmt.Sprintf("[%s]", strings.Replace(cpStr, ".", ",", -1))
+	cpMap["consented_providers"] = json.RawMessage(cpStr)
+
+	//
+	cpJSON, _ := json.Marshal(cpMap)
+	userExtMap["consented_providers_settings"] = cpJSON
+
+	extJson, extErr := json.Marshal(userExtMap)
+	if extErr != nil {
+		return request
+	}
+
+	// Start request closing
+	reqJSON, err = json.Marshal(request)
+	if err != nil {
+		return request
+	}
+
+	var reqClone openrtb2.BidRequest
+	if err = json.Unmarshal(reqJSON, &reqClone); err != nil {
+		return request
+	}
+	// End request closing
+
+	reqClone.User.Ext = extJson
+
+	return reqClone
+}
+
+func hasAdditionalConsent(consentStr string) (string, error) {
```

Comment:
this function is redundant: it is called only once and the body is just another if-then logic as most of the other code inside applyAdditionalConsentProviders so please remove the function and move the code inside applyAdditionalConsentProviders.